velox/exec/tests/ArrowStreamTest.cpp

```diff
@@ -0,0 +1,241 @@
+          : reader_(std::move(reader)) {}
+
+      std::shared_ptr<ArrowReader> reader_;
+      std::string last_error_;
+
+      PrivateData() = default;
+      DISALLOW_COPY_AND_ASSIGN(PrivateData);
+    };
+
+    explicit ExportedArrayStream(struct ArrowArrayStream* stream)
+        : stream_(stream) {}
+
+    int GetSchema(struct ArrowSchema* outSchema) {
+      reader()->GetArrowSchema(*outSchema);
+      return 0;
+    }
+
+    int GetNext(struct ArrowArray* outArray) {
+      return reader()->GetNext(outArray);
+    }
+
+    const char* GetLastError() {
+      const auto& lastError = private_data()->last_error_;
+      return lastError.empty() ? nullptr : lastError.c_str();
+    }
+
+    void Release() {
+      if (Helper::ArrowArrayStreamIsReleased(stream_)) {
+        return;
+      }
+      DCHECK_NE(private_data(), nullptr);
+      delete private_data();
+
+      Helper::ArrowArrayStreamMarkReleased(stream_);
+    }
+
+    static int StaticGetSchema(
+        struct ArrowArrayStream* stream,
+        struct ArrowSchema* outSchema) {
+      ExportedArrayStream self{stream};
+      return self.GetSchema(outSchema);
+    }
+
+    static int StaticGetNext(
+        struct ArrowArrayStream* stream,
+        struct ArrowArray* outArray) {
+      ExportedArrayStream self{stream};
+      return self.GetNext(outArray);
+    }
+
+    static void StaticRelease(struct ArrowArrayStream* stream) {
+      ExportedArrayStream{stream}.Release();
+    }
+
+    static const char* StaticGetLastError(struct ArrowArrayStream* stream) {
+      return ExportedArrayStream{stream}.GetLastError();
+    }
+
+   private:
+    std::shared_ptr<memory::MemoryPool> pool_;
+    struct ArrowArrayStream* stream_;
+    PrivateData* private_data() {
+      return reinterpret_cast<PrivateData*>(stream_->private_data);
+    }
+    const std::shared_ptr<ArrowReader>& reader() {
+      return private_data()->reader_;
+    }
+  };
+
+  void ExportArrowStream(
+      std::shared_ptr<ArrowReader> reader,
+      struct ArrowArrayStream* out) {
+    out->get_schema = ExportedArrayStream::StaticGetSchema;
+    out->get_next = ExportedArrayStream::StaticGetNext;
+    out->get_last_error = ExportedArrayStream::StaticGetLastError;
+    out->release = ExportedArrayStream::StaticRelease;
+    out->private_data = new ExportedArrayStream::PrivateData{std::move(reader)};
+  }
+
+  void testArrowStreamOutput(
+      const std::vector<RowVectorPtr>& vectors,
+      const RowTypePtr& type) {
+    auto reader = std::make_shared<ArrowReader>(pool_, vectors, type);
+    struct ArrowArrayStream arrowStream;
+    ExportArrowStream(reader, &arrowStream);
+    auto plan = std::make_shared<core::ArrowStreamNode>(
+        "0",
+        type,
+        std::move(std::make_shared<ArrowArrayStream>(arrowStream)),
+        pool_.get());
+    assertQuery(plan, "select * from tmp");
+  }
+
+  std::shared_ptr<memory::MemoryPool> pool_{memory::getDefaultMemoryPool()};
```

Comment:
This is already available in the base class. Let's remove.

Reply:
Fixed.